### PR TITLE
Lesson 6 Content

### PR DESCRIPTION
**Pull request by gildedgardenia**

### CODE REVIEW

File: src/chapters/inheritance/oop-inheritance.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+.. _inheritance:
+
+Inheritance in OOP
+==================
+
+.. index:: ! inheritance, ! subtyping
+
+**Inheritance** is the second of the **Four Pillars of Object-Oriented
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/oop-inheritance.rst]: since we dont do it elsewhere, i'd drop the capitalization and boldening of the four pillars

File: src/chapters/inheritance/csharp-inheritance.rst [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+``HouseCat`` is a class that inherits from ``Cat``, so ``HouseCat`` 
+receives the data and behaviors of ``Cat``. These inherited traits are things like 
+fields, properties, and methods. Any fields and non-constructor methods in ``Cat`` 
+are available to each instance of ``HouseCat``. 
+
+.. index:: ! extends
+
+When we speak about an inheritance relationship, we say that a ``HouseCat`` *is a* 
+``Cat``, or **extends** ``Cat``. In order to define a class that inherits from
+another, we use the ``:`` keyword.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/csharp-inheritance.rst]: can you rephrase this since ":" isnt a word

File: src/chapters/inheritance/csharp-inheritance.rst [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+.. sourcecode:: csharp
+   :linenos:
+
+   public class Cat {
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/csharp-inheritance.rst]: start brackets on new lines

File: src/chapters/inheritance/csharp-inheritance.rst [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+You can see that classes ``B``, ``C``, and ``D`` all extend class ``A``.
+And class ``E`` extends class ``D`` which itself extends class ``A``. So
+class ``E`` involves an even greater specialization of behavior than
+class ``D``.
+
+Fields and non-constructor methods are directly
+available to instances of the subclass, subject to any access modifiers.
+In general, this means that ``private`` and ``internal``
+members of a base class are not accessible to a subclass.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/csharp-inheritance.rst:70]: I'd put this sentence from the older text in here too - "However, if the subclass and base class are in the same assembly, ``internal`` allows access to a member."

File: src/chapters/inheritance/cat-classes.rst [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+Notice that ``Cat`` has a property called ``Family``, representing
+the biological family of all cats. 
+
+.. sourcecode:: csharp
+   :lineno-start: 10
+
+   public string Family { get; } = "Felidae";
+
+The ``family`` field is not directly accessible by ``HouseCat``.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: if this is public, is it not accessible by HouseCat?

File: src/chapters/inheritance/cat-classes.rst [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+However, it may be read via the getter. There is no setter for ``Family``, however.
+It may only be changed within ``Cat``. 
+It makes sense that a subclass should not be able to change the biological family 
+of a cat, since this property should rarely, if ever, change.
+
+Methods of the base class ``Cat`` may be called on instances of the
+subclass ``HouseCat`` as if they were defined as part of the
+``HouseCat``.
+
+Try it out. In your project, open up ``Program.cs``. Within the ``Main`` method
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: Within the ``Main`` method,

File: src/chapters/inheritance/cat-classes.rst [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+   :linenos:
+
+   HouseCat garfield = new HouseCat("Garfield", 12.0);
+   garfield.Eat();
+   Console.WriteLine(garfield.Tired);   // prints true
+
+The ``Eat`` method was defined in ``Cat``, but may be called on all
+``HouseCat`` instances as well. We say: “``HouseCat`` inherits the
+method ``Eat()`` from ``Cat``.” We know we have successfully called ``Eat()`` on 
+``garfield`` because the printed statement indicates the cat is now tired.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst:40]: good! very clear section here

File: src/chapters/inheritance/cat-classes.rst [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+.. sourcecode:: csharp
+   :lineno-start: 11
+
+   public HouseCat(string name, double weight) : base(weight)
+   {
+      Name = name;
+   }
+
+Here, the ``base`` keyword allows the subclass to access the constructor from the
+base class. This call to the base class constructor must be on the constructor signature.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: how about something like,

here, we use the colon again, along with the base keyword ...

i think we need to highlight the whole construction not just base.
I'd also put in a plug for thinking of this like the extension we talk about above. something like,
you can think of this as reading like: this constructor extends the base constructor that takes this weight parameter...
or something, you get it?

Comment (CarlyLanglois) [reply src/chapters/inheritance/cat-classes.rst]: also i'd modify this to say: This call to the base class constructor must be on the sybtype's constructor signature.

File: src/chapters/inheritance/cat-classes.rst [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+   :lineno-start: 16
+
+   public HouseCat(String name)
+   {
+      Name = name;
+   }
+
+Even though we don’t explicitly specify that we want to call a
+constructor from ``Cat``, the no-argument constructor will be called. Now, we can 
+initialize a new ``HouseCat`` with only a ``Name`` property and the ``Cat`` no-argument
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: Now, we can 
initialize a new ``HouseCat`` with only a name field provided and the ``Cat`` no-argument ...

File: src/chapters/inheritance/cat-classes.rst [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+class constructor has been called:
+
+.. sourcecode:: csharp
+   :linenos:
+
+   HouseCat spike = new HouseCat("Spike");
+   Console.WriteLine(spike.Weight);   // prints 13
+
+As a consequence of this constructor syntax, we can easily expose any
+constructor from the base class by providing a subclass constructor that calls the base constructor in the signature and has an empty body.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: As a consequence of this constructor syntax, we can easily expose any
constructor from the base class by providing a subclass constructor that has the same signature, no method body, and calls the base class constructor with ``: base``.

File: src/chapters/inheritance/cat-classes.rst [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+------------
+
+Sometimes when extending a class, we’ll want to modify behavior provided
+by the base class. This can be done by replacing the implementation of
+an inherited method by a completely new method implementation. For a
+given method, we can do this via **method overriding**.
+
+In our example, the ``Noise`` method of ``HouseCat`` overrides the
+method of the same name in ``Cat``. When we override it, we should use
+the ``override`` in the signature of the method in the subclass and ``virtual`` in the signature of the base class.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: When we override it, we  use ``override`` in the signature of the method in the subclass and ``virtual`` in the signature of the base class.

File: src/chapters/inheritance/cat-classes.rst [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+   }
+
+In ``HouseCat``:
+
+.. sourcecode:: csharp
+   :lineno-start: 26
+
+   public override string Noise()
+   {
+      return "Hello, my name is " + name + "!";
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: Name

File: src/chapters/inheritance/cat-classes.rst [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+   }
+
+If we have a ``HouseCat`` object and call its ``Noise()`` method, we will be 
+using the method defined in ``HouseCat``.
+
+.. sourcecode:: csharp
+   :linenos:
+
+   Cat plainCat = new Cat(8.6);
+   HouseCat cheshireCat = new HouseCat("Cheshire Caat", 26.0);
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: Cat*

File: src/chapters/inheritance/cat-classes.rst [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+
+   .. sourcecode:: csharp
+
+      public string Noise();
+
+When overriding a method, we may call the method from the base class
+that we are overriding by using ``base``. Modify your ``HouseCat.Noise()``
+method as follows:
+
+.. sourcecode:: csharp
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst:206]: couple of typos in this directive

File: src/chapters/inheritance/cat-classes.rst [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+``Object`` Class
+----------------
+
+In a previous lesson, we introduced the :ref:`special-methods` ``Equals`` and
+``ToString``. All classes contain default implementations of these methods that 
+can be overridden.
+
+In fact, these default methods are part of a class called ``Object``. If
+a class does not explicitly extend another class, then it implicitly
+extends ``Object``. So the default implementations of ``Equals`` and
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: All classes implicitly extend Object.

File: src/chapters/inheritance/cat-classes.rst [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+``ToString`` (along with a few `other
+methods <https://docs.microsoft.com/en-us/dotnet/api/system.object?view=netframework-4.8>`__)
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: https://docs.microsoft.com/en-us/dotnet/api/system.object?view=netframework-4.8#methods

File: src/chapters/inheritance/cat-classes.rst [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+   a. ``Friendly``
+   b. ``Language`` 
+   c. ``Text``
+   d. ``Message``
+   e. ``friendly``
+   f. ``language``
+   g. ``text``
+   
+
+.. ans: ``Friendly``, ``Text``, ``Language``, and ``Message``
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: the subclass will inherit the fields and properties and not the constructors. there's another level of complexity in c# that im not sure exists in java - visibility? https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/inheritance

you might want to test this out to confirm if the private backing fields will be available and visible in a subclass

File: src/chapters/inheritance/abstract-inheritance.rst [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+abstract method, we force any class that extends ``Cat`` to provide its
+own implementation of that behavior.
+
+.. sourcecode:: csharp
+
+   public abstract string Noise();
+
+
+Now, classes such as ``HouseCat`` and ``Tiger``, which both extend
+``Cat``, *must* provide their own version of ``Noise()``, with the exact
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/abstract-inheritance.rst]: drop the parens since you dont use them elsewhere on the page

File: src/chapters/inheritance/abstract-inheritance.rst [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+same method signature.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/abstract-inheritance.rst]: id note the same signature but without ``abstract``

File: src/chapters/inheritance/casting.rst [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+
+   // As long as suki really is a HouseCat, this works
+   (suki as HouseCat).IsSatisfied();
+
+The danger here is that if ``suki`` is in fact not a ``HouseCat`` (it
+was declared only as a ``Cat``, after all) then we’ll experience a
+runtime exception. A **runtime exception** is an error that occurs upon
+running the program, and is not found by the compiler beforehand. These
+are dangerous, and situations where they might come up should be
+avoided. So you should only cast an object to another type when you are
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/casting.rst]: we just said this is danger so i'd cut this sentence to: 
situations where these exeptions might come up should be
avoided.

File: src/chapters/inheritance/casting.rst [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+very sure that it’s safe to do so.
+
+Storing objects of one type (e.g. ``HouseCat``) in a variable or field
+of another “compatible” type (e.g. ``Cat``) is an example of
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/casting.rst]: drop the quotation marks on compatible

File: src/chapters/inheritance/casting.rst [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+**polymorphism**. Polymorphism is another one of the pillars of OOP and we’ll 
+have more to say about it in a future lesson.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/casting.rst]: in the next lesson

File: src/chapters/inheritance/casting.rst [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+      {
+         private bool waving;
+         
+         public Greeting(string language, string text) : base(language, text)
+         {
+         }
+
+         public void Wave()
+         {
+            waving = true;
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/casting.rst]: you need an accessor to do this without ``this``

File: src/chapters/inheritance/testing-inheritance.rst [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+
+   [TestMethod]
+   public void InheritsBaseInFirstConstructor()
+   {
+      HouseCat keyboardCat = new HouseCat("Keyboard Cat", 7);
+      Assert.AreEqual(7, keyboardCat.Weight, .001);
+   }
+
+Here, we're testing that one of our ``HouseCat`` constructors will call the ``Cat`` constructor
+and appropriately assign the ``HouseCat`` object's ``Weight`` property. Remember, we don't need
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/testing-inheritance.rst]: weight field

File: src/chapters/inheritance/testing-inheritance.rst [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+
+   a. ``Assert.IsFalse(keyboardCat.Tired);``
+
+   b. ``Assert.IsTrue(keyboardCat.Tired);``
+
+   c. ``Assert.IsTrue(keyboardCat.Hungry);``
+
+   d. ``Assert.IsFalse(keyboardCat.tired);``
+
+.. ans b, ``Assert.IsTrue(keyboardCat.Tired);``
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/testing-inheritance.rst:96]: did that description of the tired/hungry logic make it into the text already? if not, can you add it to the demo repo to help clarify the relationship between these two fields and the eating action?

File: src/chapters/inheritance/testing-inheritance.rst [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+and more to validate that the subclass constructor has inherited the base class constructor.
+
+It's a good practice to test your subclasses to verify the items that they inherit or override.
+
+Check Your Understanding
+------------------------
+
+.. admonition:: Question
+
+   Fill in the blank to test that the default constructor of ``Cat`` is called when the second
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/testing-inheritance.rst]: chris cautioned me once against using default constructor when referring to the no-arg constructor - there is apparently a more basic default cons. on all objects

File: src/chapters/inheritance/oop-inheritance.rst (hunk 1 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/oop-inheritance.rst]: Done!

File: src/chapters/inheritance/csharp-inheritance.rst (hunk 2 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/csharp-inheritance.rst]: Done!

File: src/chapters/inheritance/csharp-inheritance.rst (hunk 3 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/csharp-inheritance.rst]: Done!

File: src/chapters/inheritance/csharp-inheritance.rst (hunk 4 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/csharp-inheritance.rst:70]: Got it!

File: src/chapters/inheritance/cat-classes.rst (hunk 6 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 7 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst:40]: Thanks!

File: src/chapters/inheritance/cat-classes.rst (hunk 9 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 10 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 11 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 12 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 13 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 14 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst:206]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 16 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!

File: src/chapters/inheritance/abstract-inheritance.rst (hunk 18 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/abstract-inheritance.rst]: Done!

File: src/chapters/inheritance/abstract-inheritance.rst (hunk 19 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/abstract-inheritance.rst]: Gotcha!

File: src/chapters/inheritance/casting.rst (hunk 20 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/casting.rst]: Done!

File: src/chapters/inheritance/casting.rst (hunk 21 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/casting.rst]: Fixed!

File: src/chapters/inheritance/casting.rst (hunk 22 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/casting.rst]: Done!

File: src/chapters/inheritance/testing-inheritance.rst (hunk 24 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/testing-inheritance.rst]: Done!

File: src/chapters/inheritance/testing-inheritance.rst (hunk 26 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/testing-inheritance.rst]: Done!

File: src/chapters/inheritance/cat-classes.rst (hunk 5 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: I clarified this paragraph a bit more, so now it states that the ``get`` accessor of the property is what grabs this value.

File: src/chapters/inheritance/cat-classes.rst (hunk 8 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: I think I got it!

File: src/chapters/inheritance/cat-classes.rst (hunk 15 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Added the ``HouseCat`` and ``Cat`` example to clarify how that happens.

File: src/chapters/inheritance/casting.rst (hunk 23 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/casting.rst]: Oops! Fixed!

File: src/chapters/inheritance/testing-inheritance.rst (hunk 25 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/testing-inheritance.rst:96]: It is in the text already

File: src/chapters/inheritance/cat-classes.rst (hunk 17 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: You can access through the properties, like ``Friendly``, but you cannot access the private field directly. I clarified a paragraph earlier in the chapter on the subject, so hopefully, this question won't feel like a trick question to students!

File: src/chapters/inheritance/cat-classes.rst [hunk 27]
```diff
@@ -209,11 +206,15 @@ method as follows:
...
    :lineno-start: 38
 
-   public override string Noise() {
-      if (isSatisfied()) {
-         return "Hello, my name is " + name + "!";
-      } else {
-         return base.Noise(); // prints "Meeeeeeooooowww!"
+   public override string Noise()
+   {
+      if (isSatisfied())
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: IsSatisfied

File: src/chapters/inheritance/cat-classes.rst [hunk 28]
```diff
@@ -230,11 +231,12 @@ In a previous lesson, we introduced the :ref:`special-methods` ``Equals`` and
...
 
-In fact, these default methods are part of a class called ``Object``. If
-a class does not explicitly extend another class, then it implicitly
-extends ``Object``. So the default implementations of ``Equals`` and
-``ToString`` (along with a few `other
-methods <https://docs.microsoft.com/en-us/dotnet/api/system.object?view=netframework-4.8>`__)
+In fact, these default methods are part of a class called ``Object``.
+If a class does not explicitly extend another class, then it implicitly extends ``Object``.
+In the case of ``Cat`` and ``HouseCat``, ``HouseCat`` explicitly extends ``Cat`` and ``Cat`` implicitly extends ``Object``.
+This means that both ``Cat`` and ``HouseCat`` can access different methods and members of the ``Object`` class.
```
Comment (CarlyLanglois) [inline src/chapters/inheritance/cat-classes.rst]: yeah this is helpful! I also just don't like this sentence: "If a class does not explicitly extend another class, then it implicitly extends ``Object``." 
Because all classes implicitly inherit from Object - whether or not they also explicitly inherit another class.
I'd just ditch that sentence and edit the following one to something like:
"In the case of ``Cat`` and ``HouseCat``, ``HouseCat`` explicitly extends ``Cat`` and both ``HouseCat`` and ``Cat`` implicitly extend ``Object``."

**Review by CarlyLanglois — APPROVED** [review]

File: src/chapters/inheritance/cat-classes.rst (hunk 27 above)
Comment (gildedgardenia) [reply src/chapters/inheritance/cat-classes.rst]: Done!
